File: tests/unit/services/concurrency_service_test.go
// Package services_test validates the TypeScript source files for TASK-4592.
// These are integration-style file-system tests that verify the types,
// models, and concurrency service files contain all required definitions.
//
// Tests WILL FAIL until the implementation files are created — this is the
// intended TDD behaviour.
package services_test

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// repoRoot resolves the repository root from the current working directory.
// Tests are run with `go test ./...` from the repo root, so we walk up to
// find go.mod.
func repoRoot(t *testing.T) string {
	t.Helper()
	dir, err := os.Getwd()
	if err != nil {
		t.Fatalf("failed to get working directory: %v", err)
	}
	for {
		if _, err := os.Stat(filepath.Join(dir, "go.mod")); err == nil {
			return dir
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			t.Fatalf("could not find repo root (no go.mod found)")
		}
		dir = parent
	}
}

// readFile reads a file relative to the repo root or fails the test.
func readFile(t *testing.T, relPath string) string {
	t.Helper()
	path := filepath.Join(repoRoot(t), relPath)
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read %s: %v", relPath, err)
	}
	return string(data)
}

// fileExists returns true if the file exists at the given repo-root-relative path.
func fileExists(t *testing.T, relPath string) bool {
	t.Helper()
	path := filepath.Join(repoRoot(t), relPath)
	_, err := os.Stat(path)
	return !os.IsNotExist(err)
}

// ── File existence ────────────────────────────────────────────────────────────

func TestValidationRunTypesFileExists(t *testing.T) {
	if !fileExists(t, "backend/src/types/validation-run.types.ts") {
		t.Fatal("backend/src/types/validation-run.types.ts does not exist")
	}
}

func TestValidationRunModelFileExists(t *testing.T) {
	if !fileExists(t, "backend/src/models/validation-run.model.ts") {
		t.Fatal("backend/src/models/validation-run.model.ts does not exist")
	}
}

func TestRunStageModelFileExists(t *testing.T) {
	if !fileExists(t, "backend/src/models/run-stage.model.ts") {
		t.Fatal("backend/src/models/run-stage.model.ts does not exist")
	}
}

func TestConcurrencyServiceFileExists(t *testing.T) {
	if !fileExists(t, "backend/src/services/concurrency.service.ts") {
		t.Fatal("backend/src/services/concurrency.service.ts does not exist")
	}
}

func TestConcurrencyServiceTestFileExists(t *testing.T) {
	if !fileExists(t, "backend/tests/unit/services/concurrency.service.test.ts") {
		t.Fatal("backend/tests/unit/services/concurrency.service.test.ts does not exist")
	}
}

// ── validation-run.types.ts ───────────────────────────────────────────────────

func TestValidationRunStatusEnumDefined(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	if !strings.Contains(src, "ValidationRunStatus") {
		t.Error("validation-run.types.ts must define ValidationRunStatus enum")
	}
}

func TestValidationRunStatusEnumMembers(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	for _, member := range []string{"PENDING", "RUNNING", "SUCCESS", "FAILED", "CANCELLED"} {
		if !strings.Contains(src, member) {
			t.Errorf("ValidationRunStatus must include member %s", member)
		}
	}
}

func TestStageTypeEnumDefined(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	if !strings.Contains(src, "StageType") {
		t.Error("validation-run.types.ts must define StageType enum")
	}
}

func TestStageTypeEnumMembers(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	for _, member := range []string{"SETUP", "LINT", "TEST", "BUILD", "DEPLOY"} {
		if !strings.Contains(src, member) {
			t.Errorf("StageType enum must include member %s", member)
		}
	}
}

func TestStageStatusEnumDefined(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	if !strings.Contains(src, "StageStatus") {
		t.Error("validation-run.types.ts must define StageStatus enum")
	}
}

func TestStageStatusEnumMembers(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	for _, member := range []string{"PENDING", "RUNNING", "SUCCESS", "FAILED", "CANCELLED"} {
		if !strings.Contains(src, member) {
			t.Errorf("StageStatus enum must include member %s", member)
		}
	}
}

func TestStageOrderConstantDefined(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	if !strings.Contains(src, "STAGE_ORDER") {
		t.Error("validation-run.types.ts must export STAGE_ORDER constant")
	}
}

func TestCreateValidationRunInputInterfaceDefined(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	if !strings.Contains(src, "CreateValidationRunInput") {
		t.Error("validation-run.types.ts must export CreateValidationRunInput interface")
	}
}

func TestCreateValidationRunInputHasRequiredFields(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	for _, field := range []string{"branchName", "commitSha", "userId", "config"} {
		if !strings.Contains(src, field) {
			t.Errorf("CreateValidationRunInput must include field %s", field)
		}
	}
}

func TestRunConfigInterfaceDefined(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	if !strings.Contains(src, "RunConfig") {
		t.Error("validation-run.types.ts must export RunConfig interface")
	}
}

func TestValidationRunRecordInterfaceDefined(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	if !strings.Contains(src, "ValidationRunRecord") {
		t.Error("validation-run.types.ts must export ValidationRunRecord interface")
	}
}

func TestValidationRunRecordHasRequiredFields(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	for _, field := range []string{"id", "branchName", "commitSha", "status", "userId", "config", "createdAt", "updatedAt", "completedAt", "cancelledAt"} {
		if !strings.Contains(src, field) {
			t.Errorf("ValidationRunRecord must include field %s", field)
		}
	}
}

func TestRunStageRecordInterfaceDefined(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	if !strings.Contains(src, "RunStageRecord") {
		t.Error("validation-run.types.ts must export RunStageRecord interface")
	}
}

func TestRunStageRecordHasRequiredFields(t *testing.T) {
	src := readFile(t, "backend/src/types/validation-run.types.ts")
	for _, field := range []string{"id", "runId", "stageName", "stageIndex", "status", "logs", "startedAt", "completedAt"} {
		if !strings.Contains(src, field) {
			t.Errorf("RunStageRecord must include field %s", field)
		}
	}
}

// ── validation-run.model.ts ───────────────────────────────────────────────────

func TestValidationRunModelImportsPrisma(t *testing.T) {
	src := readFile(t, "backend/src/models/validation-run.model.ts")
	if !strings.Contains(src, "prisma") {
		t.Error("validation-run.model.ts must import/use the prisma client")
	}
}

func TestValidationRunModelCreateMethod(t *testing.T) {
	src := readFile(t, "backend/src/models/validation-run.model.ts")
	if !strings.Contains(src, "create") {
		t.Error("validation-run.model.ts must export a create method")
	}
}

func TestValidationRunModelFindByIdMethod(t *testing.T) {
	src := readFile(t, "backend/src/models/validation-run.model.ts")
	if !strings.Contains(src, "findById") {
		t.Error("validation-run.model.ts must export a findById method")
	}
}

func TestValidationRunModelFindByIdWithStagesMethod(t *testing.T) {
	src := readFile(t, "backend/src/models/validation-run.model.ts")
	if !strings.Contains(src, "findByIdWithStages") {
		t.Error("validation-run.model.ts must export a findByIdWithStages method")
	}
}

func TestValidationRunModelFindActiveOnBranchMethod(t *testing.T) {
	src := readFile(t, "backend/src/models/validation-run.model.ts")
	if !strings.Contains(src, "findActiveOnBranch") {
		t.Error("validation-run.model.ts must export a findActiveOnBranch method")
	}
}

func TestValidationRunModelUpdateStatusMethod(t *testing.T) {
	src := readFile(t, "backend/src/models/validation-run.model.ts")
	if !strings.Contains(src, "updateStatus") {
		t.Error("validation-run.model.ts must export an updateStatus method")
	}
}

func TestValidationRunModelListMethod(t *testing.T) {
	src := readFile(t, "backend/src/models/validation-run.model.ts")
	if !strings.Contains(src, "list") {
		t.Error("validation-run.model.ts must export a list method")
	}
}

func TestValidationRunModelFindActiveOnBranchChecksPendingStatus(t *testing.T) {
	src := readFile(t, "backend/src/models/validation-run.model.ts")
	if !strings.Contains(src, "PENDING") {
		t.Error("validation-run.model.ts findActiveOnBranch must check for PENDING status")
	}
}

func TestValidationRunModelFindActiveOnBranchChecksRunningStatus(t *testing.T) {
	src := readFile(t, "backend/src/models/validation-run.model.ts")
	if !strings.Contains(src, "RUNNING") {
		t.Error("validation-run.model.ts findActiveOnBranch must check for RUNNING status")
	}
}

// ── run-stage.model.ts ────────────────────────────────────────────────────────

func TestRunStageModelImportsPrisma(t *testing.T) {
	src := readFile(t, "backend/src/models/run-stage.model.ts")
	if !strings.Contains(src, "prisma") {
		t.Error("run-stage.model.ts must import/use the prisma client")
	}
}

func TestRunStageModelCreateAllMethod(t *testing.T) {
	src := readFile(t, "backend/src/models/run-stage.model.ts")
	if !strings.Contains(src, "createAll") {
		t.Error("run-stage.model.ts must export a createAll method")
	}
}

func TestRunStageModelUpdateStageMethod(t *testing.T) {
	src := readFile(t, "backend/src/models/run-stage.model.ts")
	if !strings.Contains(src, "updateStage") {
		t.Error("run-stage.model.ts must export an updateStage method")
	}
}

func TestRunStageModelFindByRunIdMethod(t *testing.T) {
	src := readFile(t, "backend/src/models/run-stage.model.ts")
	if !strings.Contains(src, "findByRunId") {
		t.Error("run-stage.model.ts must export a findByRunId method")
	}
}

func TestRunStageModelCreateAllAcceptsRunId(t *testing.T) {
	src := readFile(t, "backend/src/models/run-stage.model.ts")
	if !strings.Contains(src, "runId") {
		t.Error("run-stage.model.ts createAll must accept a runId parameter")
	}
}

// ── concurrency.service.ts ────────────────────────────────────────────────────

func TestConcurrencyServiceImportsPrisma(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	if !strings.Contains(src, "prisma") {
		t.Error("concurrency.service.ts must import/use the prisma client")
	}
}

func TestConcurrencyServiceExportsCheckAndLockBranch(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	if !strings.Contains(src, "checkAndLockBranch") {
		t.Error("concurrency.service.ts must export checkAndLockBranch function")
	}
}

func TestConcurrencyServiceExportsCreateRun(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	if !strings.Contains(src, "createRun") {
		t.Error("concurrency.service.ts must export createRun function")
	}
}

func TestConcurrencyServiceExportsGetRunWithStages(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	if !strings.Contains(src, "getRunWithStages") {
		t.Error("concurrency.service.ts must export getRunWithStages function")
	}
}

func TestConcurrencyServiceUsesTransaction(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	if !strings.Contains(src, "$transaction") {
		t.Error("concurrency.service.ts checkAndLockBranch must use prisma.$transaction for atomicity")
	}
}

func TestConcurrencyServiceUsesRawSQLForLocking(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	hasQueryRaw := strings.Contains(src, "$queryRaw") || strings.Contains(src, "$executeRaw")
	hasAdvisoryLock := strings.Contains(src, "pg_try_advisory") || strings.Contains(src, "FOR UPDATE")
	if !hasQueryRaw && !hasAdvisoryLock {
		t.Error("concurrency.service.ts must use raw SQL ($queryRaw or advisory lock) to prevent TOCTOU race conditions")
	}
}

func TestConcurrencyServiceCheckAndLockBranchReturnShape(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	if !strings.Contains(src, "allowed") {
		t.Error("concurrency.service.ts checkAndLockBranch must return an object with 'allowed' property")
	}
}

func TestConcurrencyServiceCheckAndLockBranchReturnsConflictingRunId(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	if !strings.Contains(src, "conflictingRunId") {
		t.Error("concurrency.service.ts checkAndLockBranch must return 'conflictingRunId' when a conflict is found")
	}
}

func TestConcurrencyServiceChecksPendingStatus(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	if !strings.Contains(src, "PENDING") {
		t.Error("concurrency.service.ts must check for PENDING status when locking branch")
	}
}

func TestConcurrencyServiceChecksRunningStatus(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	if !strings.Contains(src, "RUNNING") {
		t.Error("concurrency.service.ts must check for RUNNING status when locking branch")
	}
}

func TestConcurrencyServiceCreateRunSetsPendingStatus(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	// createRun must set status to PENDING when creating a new run
	if !strings.Contains(src, "PENDING") {
		t.Error("concurrency.service.ts createRun must set status to PENDING")
	}
}

func TestConcurrencyServiceGetRunWithStagesSortsByStageIndex(t *testing.T) {
	src := readFile(t, "backend/src/services/concurrency.service.ts")
	if !strings.Contains(src, "stageIndex") {
		t.Error("concurrency.service.ts getRunWithStages must sort/order stages by stageIndex")
	}
}

File: backend/tests/unit/services/concurrency.service.test.ts
/**
 * Unit tests for concurrency.service.ts — TASK-4592
 *
 * Tests the public interface of:
 *   - checkAndLockBranch(branchName): atomically checks whether a new validation
 *     run is allowed on a branch and returns { allowed } or { allowed, conflictingRunId }
 *   - createRun(input): persists a new ValidationRun with status PENDING
 *   - getRunWithStages(id): returns the run with its stages sorted by stageIndex
 */

import {
  checkAndLockBranch,
  createRun,
  getRunWithStages,
} from '../../../src/services/concurrency.service';
import { prisma } from '../../../src/db/client';

jest.mock('../../../src/db/client');

const mockPrisma = prisma as jest.Mocked<typeof prisma>;

describe('ConcurrencyService', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  // ---------------------------------------------------------------------------
  // checkAndLockBranch
  // ---------------------------------------------------------------------------

  describe('checkAndLockBranch', () => {
    it('returns { allowed: true } when no active run exists on the branch', async () => {
      // No PENDING or RUNNING rows found — lock is available
      (mockPrisma.$transaction as jest.Mock).mockImplementation(async (fn: Function) =>
        fn(mockPrisma),
      );
      (mockPrisma.$queryRaw as jest.Mock).mockResolvedValue([]);

      const result = await checkAndLockBranch('feature/my-branch');

      expect(result.allowed).toBe(true);
      expect((result as any).conflictingRunId).toBeUndefined();
    });

    it('returns { allowed: false, conflictingRunId } when a RUNNING run exists on the branch', async () => {
      const runningRunId = 'run-abc-123';

      (mockPrisma.$transaction as jest.Mock).mockImplementation(async (fn: Function) =>
        fn(mockPrisma),
      );
      (mockPrisma.$queryRaw as jest.Mock).mockResolvedValue([
        { id: runningRunId, status: 'RUNNING', branch_name: 'feature/my-branch' },
      ]);

      const result = await checkAndLockBranch('feature/my-branch');

      expect(result.allowed).toBe(false);
      expect((result as any).conflictingRunId).toBe(runningRunId);
    });

    it('returns { allowed: false } when a PENDING run exists on the branch', async () => {
      const pendingRunId = 'run-def-456';

      (mockPrisma.$transaction as jest.Mock).mockImplementation(async (fn: Function) =>
        fn(mockPrisma),
      );
      (mockPrisma.$queryRaw as jest.Mock).mockResolvedValue([
        { id: pendingRunId, status: 'PENDING', branch_name: 'feature/my-branch' },
      ]);

      const result = await checkAndLockBranch('feature/my-branch');

      expect(result.allowed).toBe(false);
    });

    it('returns { allowed: true } when only FAILED/SUCCESS/CANCELLED runs exist on the branch', async () => {
      // Terminal-state runs do NOT block new runs
      (mockPrisma.$transaction as jest.Mock).mockImplementation(async (fn: Function) =>
        fn(mockPrisma),
      );
      // The query filters to only PENDING/RUNNING rows — so the result is empty
      (mockPrisma.$queryRaw as jest.Mock).mockResolvedValue([]);

      const result = await checkAndLockBranch('feature/my-branch');

      expect(result.allowed).toBe(true);
    });

    it('wraps the concurrency check inside a prisma.$transaction call', async () => {
      (mockPrisma.$transaction as jest.Mock).mockImplementation(async (fn: Function) =>
        fn(mockPrisma),
      );
      (mockPrisma.$queryRaw as jest.Mock).mockResolvedValue([]);

      await checkAndLockBranch('feature/my-branch');

      expect(mockPrisma.$transaction).toHaveBeenCalled();
    });

    it('uses a raw SQL query to perform the SELECT ... FOR UPDATE SKIP LOCKED check', async () => {
      (mockPrisma.$transaction as jest.Mock).mockImplementation(async (fn: Function) =>
        fn(mockPrisma),
      );
      (mockPrisma.$queryRaw as jest.Mock).mockResolvedValue([]);

      await checkAndLockBranch('feature/my-branch');

      expect(mockPrisma.$queryRaw).toHaveBeenCalled();
    });
  });

  // ---------------------------------------------------------------------------
  // createRun
  // ---------------------------------------------------------------------------

  describe('createRun', () => {
    it('persists a ValidationRun with status PENDING and all provided fields', async () => {
      const input = {
        branchName: 'feature/my-branch',
        commitSha: 'abc1234567890',
        userId: 'user-001',
        config: { stages: ['SETUP', 'LINT', 'TEST'] },
      };
      const persistedRun = {
        id: 'run-new-001',
        ...input,
        status: 'PENDING',
        createdAt: new Date(),
        updatedAt: new Date(),
        completedAt: null,
        cancelledAt: null,
      };

      (mockPrisma.validationRun.create as jest.Mock).mockResolvedValue(persistedRun);

      const result = await createRun(input);

      expect(result.status).toBe('PENDING');
      expect(result.branchName).toBe(input.branchName);
      expect(result.commitSha).toBe(input.commitSha);
      expect(result.userId).toBe(input.userId);
      expect(result.config).toEqual(input.config);
    });

    it('calls prisma.validationRun.create with data containing status PENDING', async () => {
      const input = {
        branchName: 'main',
        commitSha: 'deadbeef',
        userId: 'user-002',
        config: {},
      };
      const persistedRun = {
        id: 'run-new-002',
        ...input,
        status: 'PENDING',
        createdAt: new Date(),
        updatedAt: new Date(),
        completedAt: null,
        cancelledAt: null,
      };

      (mockPrisma.validationRun.create as jest.Mock).mockResolvedValue(persistedRun);

      await createRun(input);

      expect(mockPrisma.validationRun.create).toHaveBeenCalledWith(
        expect.objectContaining({
          data: expect.objectContaining({ status: 'PENDING' }),
        }),
      );
    });

    it('returns the newly created run record with an id', async () => {
      const input = {
        branchName: 'feature/new-feature',
        commitSha: 'cafebabe',
        userId: 'user-003',
        config: { stages: ['BUILD'] },
      };
      const persistedRun = {
        id: 'run-new-003',
        ...input,
        status: 'PENDING',
        createdAt: new Date(),
        updatedAt: new Date(),
        completedAt: null,
        cancelledAt: null,
      };

      (mockPrisma.validationRun.create as jest.Mock).mockResolvedValue(persistedRun);

      const result = await createRun(input);

      expect(result.id).toBeDefined();
      expect(typeof result.id).toBe('string');
    });
  });

  // ---------------------------------------------------------------------------
  // getRunWithStages
  // ---------------------------------------------------------------------------

  describe('getRunWithStages', () => {
    it('returns run with stages sorted by stageIndex ascending', async () => {
      const runId = 'run-with-stages-001';
      // Stages deliberately out of order to confirm sorting
      const unsortedStages = [
        { id: 'stage-c', stageIndex: 2, stageName: 'TEST', status: 'PENDING', logs: '', startedAt: null, completedAt: null },
        { id: 'stage-a', stageIndex: 0, stageName: 'SETUP', status: 'PENDING', logs: '', startedAt: null, completedAt: null },
        { id: 'stage-b', stageIndex: 1, stageName: 'LINT', status: 'PENDING', logs: '', startedAt: null, completedAt: null },
      ];
      const run = {
        id: runId,
        branchName: 'feature/sorted-stages',
        commitSha: 'abc',
        status: 'RUNNING',
        config: {},
        userId: 'user-001',
        createdAt: new Date(),
        updatedAt: new Date(),
        completedAt: null,
        cancelledAt: null,
        stages: unsortedStages,
      };

      (mockPrisma.validationRun.findUnique as jest.Mock).mockResolvedValue(run);

      const result = await getRunWithStages(runId);

      expect(result).not.toBeNull();
      expect(result!.stages).toHaveLength(3);
      expect(result!.stages[0].stageIndex).toBe(0);
      expect(result!.stages[1].stageIndex).toBe(1);
      expect(result!.stages[2].stageIndex).toBe(2);
    });

    it('returns null when the run does not exist', async () => {
      (mockPrisma.validationRun.findUnique as jest.Mock).mockResolvedValue(null);

      const result = await getRunWithStages('non-existent-run-id');

      expect(result).toBeNull();
    });

    it('returns run with an empty stages array when the run has no stages', async () => {
      const runId = 'run-no-stages';
      const run = {
        id: runId,
        branchName: 'feature/no-stages',
        commitSha: 'def',
        status: 'PENDING',
        config: {},
        userId: 'user-001',
        createdAt: new Date(),
        updatedAt: new Date(),
        completedAt: null,
        cancelledAt: null,
        stages: [],
      };

      (mockPrisma.validationRun.findUnique as jest.Mock).mockResolvedValue(run);

      const result = await getRunWithStages(runId);

      expect(result).not.toBeNull();
      expect(result!.stages).toHaveLength(0);
    });

    it('queries using the provided run id', async () => {
      const runId = 'run-query-check';

      (mockPrisma.validationRun.findUnique as jest.Mock).mockResolvedValue(null);

      await getRunWithStages(runId);

      expect(mockPrisma.validationRun.findUnique).toHaveBeenCalledWith(
        expect.objectContaining({
          where: expect.objectContaining({ id: runId }),
        }),
      );
    });

    it('includes stages relation in the query', async () => {
      const runId = 'run-include-stages';

      (mockPrisma.validationRun.findUnique as jest.Mock).mockResolvedValue(null);

      await getRunWithStages(runId);

      expect(mockPrisma.validationRun.findUnique).toHaveBeenCalledWith(
        expect.objectContaining({
          include: expect.objectContaining({ stages: expect.anything() }),
        }),
      );
    });
  });
});
